dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+   2   one = 1
+   3   two = one[i - 1]
+   4   one = 2
+   5 }
+
+translates to
+
+::
+
+   MultiStage_0 [parallel]
+       {
+         Stage_0
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 1;
+           }
+           Extents: [(-1,0),(0,0),(0,0)]
+         }
+         Stage_1
+         {
+           Do_0 { Start : End }
+           {
+             two[<no_horizontal_offset>,0] = one[-1,0,0];
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+         Stage_2
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 2;
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+       }
+
+--------------
+
+Compute Domain
+~~~~~~~~~~~~~~
+
+If the vertical region does not specify any horizontal extent, dawn
+ensures that the compute-domain of statements that others depend on is
+extended to fill all the intermediate fields as required. If a
+horizontal compute-domain is specified, it is never extended.
+
+Internal Contract of Data Structures
+------------------------------------
+
+StencilInstantiation
+~~~~~~~~~~~~~~~~~~~~
+
+StencilInstantiations (SI) are user defined and are never changed during
+any optimization. Each SI will be globally split from every other one
+and there is no optimization across SIs. For every SI, a callable API
+will be generated.
+
+Stencil
+~~~~~~~
+
+Stencils serve no purpose right now
+
+--------------
+
+   It might be useful to remove them altogether since even if they
+   existed, no concept can't be captured by the other structures
+
+--------------
+
+Multistage
+~~~~~~~~~~
+
+A Multistage (MS) is the equivalent of a global synchronization between
+all processes on the full domain. each Multistage can only have one loop
+order
+
+-  in the CUDA back end they are the equivalent of a kernel call
+-  in the GridTools back end they are the equivalent of a multistage
+-  in the naive back end every storage will be synchronized on the host
+   at the beginning of each multistage
+
+--------------
+
+   Can we discuss why we still have the loop order on this level? Would
+   it potentially be interesting to promote this to the Do level and
+   change the fusion strategy if we use the GridTools back end? Or do we
+   know that we are losing performance in the CUDA back end if we mix
+   forward and backward loops versus new kernel calls?
+
+--------------
+
+Stage
+~~~~~
+
+A Stage (St) specifies a specific compute-domain for different
+computations. These can be either global computation bounds or extended
+compute domains due to larger required input data.
+
+Each stage has the option to require (block-wide) synchronization.
+
+-  in the CUDA back end they are equivalent to conditionals that check
+   if the block / thread combination is inside the compute domain and a
+   potential ``__syncthreads()`` afterwards.
```

Comment:
is there a reason to have it only "potentially"? Because if you don't need synchronization you could put them into the same Do?

Reply:
yes, currently the reason is that stages cannot have overlapping do-methods. so you might need to generate two stages to fulfil that requirement even though they do not require synchronisation.

Reply:
I see. But there is nothing fundamental which prohibits us doing this, right?

Reply:
I would say no